pkg/schedule/affinity/group.go

```diff
@@ -0,0 +1,378 @@
+		RegionCount:               len(g.Regions),
+		AffinityRegionCount:       g.AffinityRegionCount,
+		affinityVer:               g.AffinityVer,
+		groupInfoPtr:              g,
+	}
+}
+
+// IsAvailable indicates that the Group is currently in the groupAvailable status,
+// which allows affinity scheduling and disallows other balancing scheduling.
+func (g *runtimeGroupInfo) IsAvailable() bool {
+	return g.GetAvailability() == groupAvailable
+}
+
+// IsExpired indicates that the Group is currently in the groupExpired status,
+// which disallows affinity scheduling and allows other balancing scheduling.
+func (g *runtimeGroupInfo) IsExpired() bool {
+	return g.GetAvailability() == groupExpired
+}
+
+// GetAvailability returns the group availability. It handles degradedExpiredAt internally.
+func (g *runtimeGroupInfo) GetAvailability() groupAvailability {
+	if g.availability == groupDegraded && uint64(time.Now().Unix()) > g.degradedExpiredAt {
+		return groupExpired
+	}
+	return g.availability
+}
+
+// SetAvailability updates availability and degradedExpiredAt with the following behavior:
+// - When the current availability is groupAvailable
+//   - newAvailability = groupAvailable: do nothing
+//   - newAvailability = groupDegraded: change the availability to groupDegraded and set degradedExpiredAt based on the current time
+//   - newAvailability = groupExpired: change the availability to groupExpired
+//
+// - When the current availability is groupDegraded
+//   - newAvailability = groupAvailable: change the availability to groupAvailable
+//   - newAvailability = groupDegraded: do nothing and do not refresh degradedExpiredAt
+//   - newAvailability = groupExpired: change the availability to groupExpired
+//
+// - When the current availability is groupExpired
+//   - newAvailability = groupAvailable: change the availability to groupAvailable
+//   - newAvailability = groupDegraded: do nothing and do not refresh degradedExpiredAt
+//   - newAvailability = groupExpired: do nothing
+func (g *runtimeGroupInfo) SetAvailability(newAvailability groupAvailability) {
+	// If the expiration time has been reached, change groupDegraded to groupExpired.
+	if g.availability == groupDegraded && g.IsExpired() {
+		g.availability = groupExpired
+	}
+	// Update availability
+	if newAvailability == groupDegraded {
+		// Only set the expiration time when transitioning from groupAvailable to groupDegraded.
+		// Do nothing if the original availability is already groupDegraded or groupExpired.
+		if g.availability == groupAvailable {
+			g.availability = groupDegraded
+			g.degradedExpiredAt = newDegradedExpiredAtFromNow()
+		}
+	} else {
+		g.availability = newAvailability
+	}
+}
+
+func newDegradedExpiredAtFromNow() uint64 {
+	return uint64(time.Now().Unix()) + defaultDegradedExpirationSeconds
+}
+
+// IsAffinitySchedulingEnabled indicates whether affinity scheduling is allowed.
+func (g *runtimeGroupInfo) IsAffinitySchedulingEnabled() bool {
```

Comment:
The name is the same as IsAffinitySchedulingEnabled in the config, but the meaning is different.